extensions/amp-a4a/0.1/real-time-config-manager.js

```diff
@@ -0,0 +1,202 @@
+export const RTC_ERROR_ENUM = {
+  BAD_JSON_RESPONSE: 'bad_json_response',
+  DUPLICATE_URL: 'duplicate_url',
+  INSECURE_URL: 'insecure_url',
+  MAX_CALLOUTS_EXCEEDED: 'max_callouts_exceeded',
+  NETWORK_FAILURE: 'network_failure',
+  UNKNOWN_VENDOR: 'unknown_vendor',
+};
+
+/**
+ * @param {!Array<Promise>} promiseArray
+ * @param {!string} error
+ * @param {!string} callout
+ * @return
+ */
+function logAndAddErrorResponse(promiseArray, error, callout) {
+  dev().warn(TAG, `Dropping RTC Callout to ${callout} due to ${error}`);
+  promiseArray.push(buildErrorResponse(error, callout));
+}
+
+/**
+ * @param {!string} error
+ * @param {!string} callout
+ * @param {number=} opt_rtcTime
+ * @return
+ */
+function buildErrorResponse(error, callout, opt_rtcTime) {
+  return {error, callout, rtcTime: opt_rtcTime || 0};
+}
+
+/**
+ * For a given A4A Element, sends out Real Time Config requests to
+ * any urls or vendors specified by the publisher.
+ * @param {!AMP.BaseElement} a4aElement
+ * @param {!Object} customMacros The ad-network specified macro
+ *   substitutions available to use.
+ * @return
+ */
+function maybeExecuteRealTimeConfig(a4aElement, customMacros) {
+  const rtcConfig = validateRtcConfig(a4aElement.element);
+  if (!rtcConfig) {
+    return;
+  }
+  const promiseArray = [];
+  const seenUrls = {};
+  const rtcStartTime = Date.now();
+  // For each publisher defined URL, inflate the url using the macros,
+  // and send the RTC request.
+  (rtcConfig['urls'] || []).forEach(url =>
+    inflateAndSendRtc_(a4aElement, url, seenUrls, promiseArray,
+        rtcStartTime, customMacros,
+        rtcConfig['timeoutMillis'])
+  );
+  // For each vendor the publisher has specified, inflate the vendor
+  // url if it exists, and send the RTC request.
+  Object.keys(rtcConfig['vendors'] || []).forEach(vendor => {
+    const url = RTC_VENDORS[vendor.toLowerCase()];
+    if (!url) {
+      return logAndAddErrorResponse(promiseArray,
+                                    RTC_ERROR_ENUM.UNKNOWN_VENDOR, vendor);
+    }
+    // The ad network defined macros override vendor defined/pub specifed.
+    const macros = Object.assign(rtcConfig['vendors'][vendor], customMacros);
+    inflateAndSendRtc_(a4aElement, url, seenUrls, promiseArray, rtcStartTime,
+        macros, rtcConfig['timeoutMillis'],
+        vendor);
+  });
+  return Promise.all(promiseArray);
+}
+
+/**
+ * @param {!AMP.BaseElement} a4aElement
+ * @param {!string} url
+ * @param {!Object<string, boolean>} seenUrls
+ * @param {!Array<Promise>} promiseArray
+ * @param {!number} rtcStartTime
+ * @param {!Object} macros
+ * @param {!number} timeoutMillis
+ * @param {string=} opt_vendor
+ * @return
+ */
+function inflateAndSendRtc_(a4aElement, url, seenUrls, promiseArray,
+                            rtcStartTime, macros, timeoutMillis,
+                            opt_vendor) {
+  const win = a4aElement.win;
+  const ampDoc = a4aElement.getAmpDoc();
+  if (seenUrls.length == MAX_RTC_CALLOUTS) {
```

Comment:
can you get the size of object this way?!  I think this needs to be Object.keys(seenUrls).length